enzyme/test/Integration/ReverseMode/invsqrt.c

```diff
@@ -66,33 +66,79 @@ float rev_rsqrt(float x, float grad_out) {
 
 void* __enzyme_register_gradient_rsqrt[3] = { (void*)Q_rsqrt, (void*)aug_rsqrt, (void*)rev_rsqrt };
 
+//// (2) That the above aug_rsqrt works is really a fluke. This is a correct interface in general.
+double rsqrt2(double x) {return Q_rsqrt(x);}
+double aug_rsqrt2(double x) {return Q_rsqrt(x);}
+double rev_rsqrt2(double x, double y_b) {return -y_b * Q_rsqrt(x) / (2 * x);}
+void* __enzyme_register_gradient_rsqrt2[3] = { (void*)rsqrt2, (void*)aug_rsqrt2, (void*)rev_rsqrt2};
+
+double invmag2(double* __restrict__ A, int n) {
+  double sumsq = 0;
+  for (int i=0; i<n; i++)
+    sumsq += A[i] * A[i];
+  return rsqrt2(sumsq);
+}
+
+//// (3) This is another possible interface for more complicated functions.
+void rsqrt3(double* y, double* x) {
+    *y = Q_rsqrt(*x);
+}
+void* aug_rsqrt3(double* y, double* y_b, double* x, double* x_b) {
+    double* tape = malloc(sizeof(double));
+    *tape = *x;
+    *y = Q_rsqrt(*x);
+    return tape;
+}
+void rev_rsqrt3(double* y, double* y_b, double* x, double* x_b, void* tape) {
+    double x0 = *(double*)tape; // x points to junk; original input *x must be obtained from the tape
+    *x_b -= (*y_b) * Q_rsqrt(x0) / (2 * x0);
+    *y_b = 0;   // Since y is used purely as an output, changes in y do not affect the calculation
+    free(tape);
```

Comment:
This is perhaps off-topic, but what if the user wishes to apply reverse mode multiple times? Should freeing the tape not be decoupled from the numerics of the reverse pass?

Reply:
If the code is generated by Enzyme, the no-free flag will keep that cache around. We need to figure out a nice interface for specifying that (and other options) for the custom derivatives.

Alternatively, perhaps we just have a special allocator that manages this behind the scenes.

Reply:
Also at least in this case, one doesn't need to return a void*, one can just use a double, rather than double* as the tape.

Reply:
I agree that in this case the tape can be of type double, but a void star will get you a lot further in more general cases. I will respond to the initial question. It is my understanding that when one writes an augmented function (f_fwd) and a reverse function (f_rev), one is writing a pair of coupled functions; every call to the augmented forward is expected to eventually followed up by precisely one corresponding call to the reverse function - and with the corresponding tapes. Given this interface, there this no other place but the reverse function to cleanup whatever mess was made by the augmented forward function. Furthermore, since the same code is run for multiple calls to f (suppose you want to differentiate a function g which calls f twice), the tape is the only way in general to distinguish the two separate pieces of data that f_rev will need.
I am not sure what this no-free flag is. However, I do know the enzyme_allocated option can handle fairly general cases - but this option is not (and cannot be?) integrated with the custom derivatives.